### PR TITLE
Add DeletePipeline API

### PR DESCRIPTION
**Pull request by xoxys**
This is just a first step, the final goal is to have an API endpoint to prune Repo Pipelines older than the given date. 

@woodpecker-ci/maintainers Can I get some feedback if this is the right direction?

### CODE REVIEW

File: server/store/datastore/pipeline.go [hunk 1]
```diff
@@ -52,9 +52,18 @@ func (s storage) GetPipelineLastBefore(repo *model.Repo, branch string, num int6
 		Get(pipeline))
 }
 
-func (s storage) GetPipelineList(repo *model.Repo, p *model.ListOptions) ([]*model.Pipeline, error) {
+func (s storage) GetPipelineList(repo *model.Repo, p *model.ListOptions, f *model.FilterOptions) ([]*model.Pipeline, error) {
 	pipelines := make([]*model.Pipeline, 0, 16)
-	return pipelines, s.paginate(p).Where("pipeline_repo_id = ?", repo.ID).
+
+	if f != nil {
+		return pipelines, s.paginate(p).Where(builder.Eq{"pipeline_repo_id": repo.ID}).
```
Comment (6543) [inline server/store/datastore/pipeline.go:59]: i would create a new condition with builder and set `builder.Eq{"pipeline_repo_id": repo.ID}` on the beginning

then have an if that will enable filters based on options:
```go
f.After != 0 
  cond = cond.And(builder.Gt{"pipeline_started": f.After}
}
```
...

Comment (6543) [reply server/store/datastore/pipeline.go:59]: PS: defining an filter type to pass into the store func is the right way yes :)

Comment (xoxys) [reply server/store/datastore/pipeline.go:59]: Thanks!

File: server/model/pipeline.go [hunk 2]
```diff
@@ -53,6 +53,11 @@ type Pipeline struct {
 	IsPrerelease        bool                    `json:"is_prerelease,omitempty"     xorm:"is_prerelease"`
 } //	@name Pipeline
 
+type FilterOptions struct {
```
Comment (6543) [inline server/model/pipeline.go:56]: PipelineFilter... ?

File: server/store/datastore/pipeline.go [hunk 3]
```diff
@@ -52,9 +52,21 @@ func (s storage) GetPipelineLastBefore(repo *model.Repo, branch string, num int6
 		Get(pipeline))
 }
 
-func (s storage) GetPipelineList(repo *model.Repo, p *model.ListOptions) ([]*model.Pipeline, error) {
+func (s storage) GetPipelineList(repo *model.Repo, p *model.ListOptions, f *model.FilterOptions) ([]*model.Pipeline, error) {
 	pipelines := make([]*model.Pipeline, 0, 16)
-	return pipelines, s.paginate(p).Where("pipeline_repo_id = ?", repo.ID).
+
+	cond := builder.NewCond()
+	cond = cond.And(builder.Eq{"pipeline_repo_id": repo.ID})
```
Comment (6543) [inline server/store/datastore/pipeline.go:59]:
```suggestion
	cond := builder.NewCond().And(builder.Eq{"pipeline_repo_id": repo.ID})
```

File: server/store/datastore/pipeline.go [hunk 4]
```diff
@@ -52,9 +52,21 @@ func (s storage) GetPipelineLastBefore(repo *model.Repo, branch string, num int6
...
+
+	if f != nil {
+		cond = cond.And(builder.Gt{"pipeline_started": f.After})
```
[the remark this answers is not in the document]
Comment (xoxys) [reply server/store/datastore/pipeline.go:61]: This will still result in a segfault as `f` can be nil

[the remark this answers is not in the document]
Comment (xoxys) [reply server/store/datastore/pipeline.go]: Adjusted

**Comment by xoxys** [comment]
Will add some more tests for the delete API later today.

**Comment by xoxys** [comment]
Should we prevent deleting active pipelines for `api.DeletePipelines`?

**Comment by 6543** [comment]
:point_up: yes & pending ... as they have to be removed from task queue

File: server/router/api.go [hunk 5]
```diff
@@ -94,6 +94,7 @@ func apiRoutes(e *gin.RouterGroup) {
 
 					repo.GET("/pipelines", api.GetPipelines)
 					repo.POST("/pipelines", session.MustPush, api.CreatePipeline)
+					repo.DELETE("/pipelines", api.DeletePipelines)
```
Comment (6543) [inline server/router/api.go:97]: must session.MustRepoAdmin()

**Comment by xoxys** [comment]
> ☝️ yes & pending ... as they have to be removed from task queue

I have added a helper function to check the status and applied it to DeletePipelineLogs as well. `model.StatusRunning, model.StatusPending, model.StatusBlocked` are blocked now, for both delete actions.

**Comment by anbraten** [comment]
@xoxys Is your plan to have something like requested in #1068? (seems to have quite a few 👍🏾)  So currently this would be done by calling the api, should we add some cron / ui button as well?

**Comment by xoxys** [comment]
> Is your plan to have something like requested in https://github.com/woodpecker-ci/woodpecker/issues/1068
In general, yes. My DB is growing constantly already.

>  should we add some cron / ui button as well?
I would like to separate this. First, let's get the API changes merged. After that, we can add UI components, CLI, etc.
Personally, I don't need a UI, I will just use the CLI to run a scheduled job to clean up old pipelines.

Note: These APIs are about deleting the entire pipeline, not only it logs.

**Comment by xoxys** [comment]
In addition to this, an API `DeletePipelineLogs` exists already and there is a PR https://github.com/woodpecker-ci/woodpecker/pull/3451 to add the ability to delete logs of particular steps.

**Comment by anbraten** [comment]
One think I am wondering is, would it be cleaner to have a delete-pipeline endpoint instead of delete-pipelines (as that could be used to delete a single pipeline) similar to delete pipeline-logs of a single pipeline etc. And then the get endpoint with time filters could be used to get all pipelines before and after x and delete them with a script etc by using their id?

**Comment by xoxys** [comment]
I would like to avoid extra scripting around it. However, I can still add a DeletePipeline API that just calls `store.DeletePipeline` to delete a single pipeline by ID. Or is there any downside to have both API endpoints?

**Comment by xoxys** [comment]
@woodpecker-ci/maintainers Anything missing for this?

**Comment by anbraten** [comment]
> I would like to avoid extra scripting around it. 

I see, but shouldn't we then integrate a cron job doing that based on a setting into wp directly then?

> However, I can also add DeletePipeline API that just calls store.DeletePipeline to delete a single pipeline by ID. Or is there any downside to have both API endpoints?

It feels quite powerful to me having an endpoint that could delete all pipelines of a repo.

**Comment by xoxys** [comment]
> It feels quite powerful to me having an endpoint that could delete all pipelines of a repo.

Please make a decision and I'll implement it that way then.

> I see, but shouldn't we then integrate a cron job doing that based on a setting into wp directly then?

Again, yes we can add client integrations (UI, CLI) later please let's get the API changes merged first.

**Comment by xoxys** [comment]
There is a `DeleteRepo` API exposed already, this will delete everything from the DB as well. To be honest, I don't see why we should not allow repo admins (they should know what they are doing) to delete all pipelines from a repo.

**Comment by xoxys** [comment]
@woodpecker-bot Can we make a decision please? I would like to move on and get it merged. A rapidly growing DB is a real issue for frequently used instances...

**Comment by xoxys** [comment]
As discussed in this thread and in https://github.com/woodpecker-ci/woodpecker/issues/1068, I would still like to have an API endpoint for deleting pipelines. Since @anbraten prefers to have an endpoint to delete a single pipeline instead of multiple, I have refactored the PR.

**Pull request merged**